### PR TITLE
feat:support set config from system.envs.

### PR DESCRIPTION
**Pull request by JaredTan95**
Please answer these questions before submitting pull request

- Why submit this pull request?
- [ ] Bug fix
- [ ] New feature provided
- [x] Improve performance

- Related issues

___
### Bug fix
- Bug description.

- How to fix?

___
### New feature or improvement
- Describe the details and related test reports.

### CODE REVIEW

**Comment by wu-sheng** [comment]
This is right doc refer, https://github.com/apache/incubator-skywalking/blob/master/docs/en/setup/backend/backend-setting-override.md

Also, the command has a typo.
`-core.default.restHost=172.0.4.12` -> `-Dcore.default.restHost=172.0.4.12`. Should fix this in your PR too.

File: oap-server/server-library/library-util/src/test/java/org/apache/skywalking/oap/server/library/util/PropertyPlaceholderHelperTest.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+                        providerConfig.forEach((name, propertiesConfig) -> {
+                            if (propertiesConfig != null) {
+                                propertiesConfig.forEach((key, value) -> properties.put(key, value));
+                            }
+                        });
+                    }
+                });
+            }
+        } catch (FileNotFoundException e) {
+            e.printStackTrace();
```
Comment (wu-sheng) [inline oap-server/server-library/library-util/src/test/java/org/apache/skywalking/oap/server/library/util/PropertyPlaceholderHelperTest.java]: Should throw exception.

File: oap-server/server-library/library-util/src/test/java/org/apache/skywalking/oap/server/library/util/PropertyPlaceholderHelperTest.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        }
+        placeholderHelper =
+            new PropertyPlaceholderHelper(PlaceholderConfigurerSupport.DEFAULT_PLACEHOLDER_PREFIX,
+                PlaceholderConfigurerSupport.DEFAULT_PLACEHOLDER_SUFFIX,
+                PlaceholderConfigurerSupport.DEFAULT_VALUE_SEPARATOR, true);
+    }
+
+    @Test
+    public void testDataType() {
```
Comment (wu-sheng) [inline oap-server/server-library/library-util/src/test/java/org/apache/skywalking/oap/server/library/util/PropertyPlaceholderHelperTest.java]: This case can't test get system properties or env value right, from my understanding. How about adding some for this?

**Comment by JaredTan95** [comment]
@wu-sheng I finished the test and doc about how to use it.

**Comment by wu-sheng** [comment]
For convenience, this is the document.

___
## System environment variables
- Example

  Override `restHost` in this setting segment through environment variables
  
```yaml
core:
  default:
    restHost: ${REST_HOST:0.0.0.0}
    restPort: 12800
    restContextPath: /
    gRPCHost: 0.0.0.
    gRPCPort: 11800
```

If the `REST_HOST ` environment variable exists in your operating system and its value is `172.0.4.12`, 
then the value of `restHost` here will be overwritten to `172.0.4.12`, otherwise, it will be set to `0.0.0.0`

**Comment by wu-sheng** [comment]
Codes look good for me, @peng-yongsheng What do you think of this? And should we put this style setting yml as our default style in startup module and in assembly?

**Comment by JaredTan95** [comment]
Placeholder nesting is also supported，like ,`${REST_HOST:${REST_URL:127.0.0.1}}` :)

**Comment by wu-sheng** [comment]
> Placeholder nesting is also supported，like ,${REST_HOST:${REST_URL:127.0.0.1}} :)

You should put this in the document.

**Comment by JaredTan95** [comment]
## System environment variables
- Example

  Override `restHost` in this setting segment through environment variables
  
```yaml
core:
  default:
    restHost: ${REST_HOST:0.0.0.0}
    restPort: 12800
    restContextPath: /
    gRPCHost: 0.0.0.
    gRPCPort: 11800
```

If the `REST_HOST ` environment variable exists in your operating system and its value is `172.0.4.12`, 
then the value of `restHost` here will be overwritten to `172.0.4.12`, otherwise, it will be set to `0.0.0.0`.

By the way, Placeholder nesting is also supported, like `${REST_HOST:${ANOTHER_REST_HOST:127.0.0.1}}`.
In this case, if the `REST_HOST ` environment variable not exists, but the ```REST_ANOTHER_REST_HOSTHOST``` 
environment variable exists and its value is `172.0.4.12`, then the value of `restHost` here will be overwritten to `172.0.4.12`, otherwise, it will be set to `127.0.0.1`.

**Comment by peng-yongsheng** [comment]
It's a good feature, I like it so much. Users can known which variables could be config clearly. And they could cover them by System.env.

**Comment by JaredTan95** [comment]
Is the environment variable name format OK? :)

```

cluster:
  standalone:
   # Please check your ZooKeeper is 3.5+, However, it is also compatible with ZooKeeper 3.4.x. Replace the ZooKeeper 3.5+
   # library the oap-libs folder with your ZooKeeper 3.4.x library.
#  zookeeper:
#    hostPort: ${SW_CLUSTER_ZK_HOST_PORT:localhost:2181}
#    #Retry Policy
#    baseSleepTimeMs: ${SW_CLUSTER_ZK_SLEEP_TIME:1000} # initial amount of time to wait between retries
#    maxRetries: ${SW_CLUSTER_ZK_MAX_RETRIES:3} # max number of times to retry
#  kubernetes:
#    watchTimeoutSeconds: ${SW_CLUSTER_K8S_WATCH_TIMEOUT:60}
#    namespace: ${SW_CLUSTER_K8S_NAMESPACE:default}
#    labelSelector: ${SW_CLUSTER_K8S_LABEL:app=collector,release=skywalking}
#    uidEnvName: ${SW_CLUSTER_K8S_UID:SKYWALKING_COLLECTOR_UID}
core:
  default:
    restHost: ${SW_CORE_REST_HOST:0.0.0.0}
    restPort: ${SW_CORE_REST_PORT:12800}
    restContextPath: ${SW_CORE_REST_CONTEXT_PATH:/}
    gRPCHost: ${SW_CORE_GRPC_HOST:0.0.0.0}
    gRPCPort: ${SW_CORE_GRPC_PORT:11800}
    downsampling:
    - Hour
    - Day
    - Month
    # Set a timeout on metric data. After the timeout has expired, the metric data will automatically be deleted.
    recordDataTTL: ${SW_CORE_RECORD_DATA_TTL:90} # Unit is minute
    minuteMetricsDataTTL: ${SW_CORE_MINUTE_METRIC_DATA_TTL:90} # Unit is minute
    hourMetricsDataTTL: ${SW_CORE_HOUR_METRIC_DATA_TTL:36} # Unit is hour
    dayMetricsDataTTL: ${SW_CORE_DAY_METRIC_DATA_TTL:45} # Unit is day
    monthMetricsDataTTL: ${SW_CORE_MONTH_METRIC_DATA_TTL:18} # Unit is month
storage:
  elasticsearch:
    clusterNodes: ${SW_STORAGE_ES_CLUSTER_NODES:localhost:9200}
    indexShardsNumber: ${SW_STORAGE_ES_INDEX_SHARDS_NUMBER:2}
    indexReplicasNumber: ${SW_STORAGE_ES_INDEX_REPLICAS_NUMBER:0}
    # Batch process setting, refer to https://www.elastic.co/guide/en/elasticsearch/client/java-api/5.5/java-docs-bulk-processor.html
    bulkActions: ${SW_STORAGE_ES_BULK_ACTIONS:2000} # Execute the bulk every 2000 requests
    bulkSize: ${SW_STORAGE_ES_BULK_SIZE:20} # flush the bulk every 20mb
    flushInterval: ${SW_STORAGE_ES_FLUSH_INTERVAL:10} # flush the bulk every 10 seconds whatever the number of requests
    concurrentRequests: ${SW_STORAGE_ES_CONCURRENT_REQUESTS:2} # the number of concurrent requests
#  h2:
#    driver: org.h2.jdbcx.JdbcDataSource
#    url: jdbc:h2:mem:skywalking-oap-db
#    user: sa
receiver-register:
  default:
receiver-trace:
  default:
    bufferPath: ${SW_RECEIVER_BUFFER_PATH:../trace-buffer/}  # Path to trace buffer files, suggest to use absolute path
    bufferOffsetMaxFileSize: ${SW_RECEIVER_BUFFER_OFFSET_MAX_FILE_SIZE:100} # Unit is MB
    bufferDataMaxFileSize: ${SW_RECEIVER_BUFFER_DATA_MAX_FILE_SIZE:500} # Unit is MB
    bufferFileCleanWhenRestart: ${SW_RECEIVER_BUFFER_FILE_CLEAN_WHEN_RESTART:false}
receiver-jvm:
  default:
service-mesh:
  default:
    bufferPath: ${SW_SERVICE_MESH_BUFFER_PATH:../mesh-buffer/}  # Path to trace buffer files, suggest to use absolute path
    bufferOffsetMaxFileSize: ${SW_SERVICE_MESH_OFFSET_MAX_FILE_SIZE:100} # Unit is MB
    bufferDataMaxFileSize: ${SW_SERVICE_MESH_BUFFER_DATA_MAX_FILE_SIZE:500} # Unit is MB
    bufferFileCleanWhenRestart: ${SW_SERVICE_MESH_BUFFER_FILE_CLEAN_WHEN_RESTART:false}
istio-telemetry:
  default:
receiver_zipkin:
  default:
    host: ${SW_RECEIVER_ZIPKIN_HOST:0.0.0.0}
    port: ${SW_RECEIVER_ZIPKIN_PORT:9411}
    contextPath: ${SW_RECEIVER_ZIPKIN_CONTEXT_PATH:/}
query:
  graphql:
    path: ${SW_QUERY_GRAPHQL_PATH:/graphql}
alarm:
  default:

```

**Comment by wu-sheng** [comment]
@JaredTan95 I am going to merge this. 

Also, consider this setting style looks good. How about moving this to agent config initial? I think this doesn't require any other dependency, right? If so, how about moving these codes into `apm-util` mdoule, and active this in both collector and agent.config? Of course this is another proposal, we should use another PR to do this, if you want.

**Comment by JaredTan95** [comment]
> @JaredTan95 I am going to merge this.
> 
> Also, consider this setting style looks good. How about moving this to agent config initial? I think this doesn't require any other dependency, right? If so, how about moving these codes into `apm-util` mdoule, and active this in both collector and agent.config? Of course this is another proposal, we should use another PR to do this, if you want.

yep, It will be considered for apply to the agent configuration.

**Comment by wu-sheng** [comment]
> yep, It will be considered for apply to the agent configuration.

I think this is more cloud native and container friendly. Thanks for bringing this good idea to the project.

**Pull request merged**